wandb/sdk/system/system_monitor.py

```diff
@@ -0,0 +1,182 @@
+import logging
+import multiprocessing as mp
+import queue
+import threading
+from typing import TYPE_CHECKING, List, Optional, Union
+
+from wandb.sdk.system.assets.asset_registry import asset_registry
+from wandb.sdk.system.assets.interfaces import Asset, Interface
+from wandb.sdk.system.system_info import SystemInfo
+
+if TYPE_CHECKING:
+    from wandb.proto.wandb_telemetry_pb2 import TelemetryRecord
+    from wandb.sdk.interface.interface import FilesDict
+    from wandb.sdk.internal.settings_static import SettingsStatic
+
+
+logger = logging.getLogger(__name__)
+
+
+class AssetInterface:
+    def __init__(self) -> None:
+        self.metrics_queue: "queue.Queue[dict]" = queue.Queue()
+        self.telemetry_queue: "queue.Queue[TelemetryRecord]" = queue.Queue()
+
+    def publish_stats(self, stats: dict) -> None:
+        self.metrics_queue.put(stats)
+
+    def _publish_telemetry(self, telemetry: "TelemetryRecord") -> None:
+        self.telemetry_queue.put(telemetry)
+
+    def publish_files(self, files_dict: "FilesDict") -> None:
+        pass
+
+
+class SystemMonitor:
+    # SystemMonitor is responsible for managing system metrics data.
+
+    # if joining assets, wait for publishing_interval times this many seconds
+    PUBLISHING_INTERVAL_DELAY_FACTOR = 2
```

Comment:
i notice this is dropping the first metric...   so we only get the first metic at time = 60 instead of time=30.

Reply:
Good point, let us explain.

A bit of context. Consider the following situation: you are monitoring two things: CPU and Apple GPU utilization. Measuring the first is fast and cheap, whereas the second takes about 1 sec on my machine:

```
wandb.init(...)  --|-----------
cpu measurement  ----|--|--|--|
gpu measurement  --------|---|-

time ->
```

Here, the publishing interval is two "-", and it takes some extra "-"s to measure gpu.

If we publish Asset metrics individually, there is no problem, only that the frontend needs to be able to digest that -- this will be the future default. 
In the compatibility mode, however, our main goal is to "catch" the first slow measurement, therefore we wait for 2 x "--", grab everything from the internal interface queue, and package the last measurements per asset into the dict that is then published to the real interface queue.
We came up with a couple alternatives with varying degree of complexity, but decided to err on the logic/code simplicity and in the future do:
- In the next PR of the series, add more accurate time stamps (save them together with the metrics and "unpack" them in the sender). When publishing per asset, this will be much more accurate, but also in the compatibility mode, the accuracy will significantly improve.
- Eventually phase out the compatibility mode :)